File: 백준/Silver/9012. 괄호/괄호.cc
#include <bits/stdc++.h>
using namespace std;

int main() {
  int t;
  cin >> t;
  while(t--) {
    string s;
    cin >> s;
    if((s.size()&1)==1 || s[0]==')') {
      cout << "NO" << endl;
    } else {
      stack<char> stk;
      for(int i=0; i<s.size(); ++i) {
        if(s[i] == '(') stk.push(s[i]);
        else {
          if(stk.empty()) {
            stk.push('X');
            break;
          } else {
            stk.pop();
          }
        }
      }
      
      if(stk.empty()) cout << "YES" << endl;
      else cout << "NO" << endl;
    }
  }
  return 0;
}

File: 백준/Silver/9012. 괄호/괄호.rb
x = gets.to_i

def vps?(str) 
	stack = []
	if str.size & 1 == 1 or str[0] == ')'
		return "NO"
	end
	str.each_char { |c|
		if c == '(' 
			stack.push(c)
		else
			if stack.empty? then return "NO" end 
			stack.pop
		end
	}
	if stack.empty? then "YES" else "NO" end
end 

x.times do 
	puts vps?(gets.chomp)
end
